### ISSUE TITLE
[docker image] [boost-exception] build failure

### ISSUE THREAD

**Issue by Nemirtingas** [microsoft/vcpkg#19202]
**Host Environment**
- OS: Linux cross-compiling for Windows.
- Compiler: clang 9 + clang-cl 9 + llvm 9

**To Reproduce**
Using docker image nemirtingas/windowscross_vcpkg:msvc2019_win10.0.18362.0

Custom triplet x86-windows-nemirtingas:
```
set(VCPKG_TARGET_ARCHITECTURE x86)
set(VCPKG_CRT_LINKAGE static)
set(VCPKG_LIBRARY_LINKAGE static)

set(VCPKG_BUILD_TYPE release)

set(ENV{HOST_ARCH} ${VCPKG_TARGET_ARCHITECTURE})

set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE /clang_windows_sdk/clang-cl-msvc.cmake)

set(ENV{VCPKG_TOOLCHAIN} "/vcpkg/scripts/toolchains/windows.cmake")
```

```
vcpkg install boost-exception:x86-windows-nemirtingas
```

**Failure logs** 
```
/vcpkg/buildtrees/boost-exception/x86-windows-nemirtingas-rel/user-config.jam:22: in modules.load
*** argument error
* rule using ( toolset-module : * )
* called with: (  :  :  : <ranlib> <archiver> <flavor>mingw )
* missing argument toolset-module
```

**Additional context**
user-config.jam seems to not be populated with the cmake's configurations variables. I believe this happens because the boost module target doesn't load my custom CMake toolchain because a lots of variables are empty:
CMAKE_SYSTEM_NAME, B2_TOOLSET, CMAKE_RANLIB, CMAKE_AR, CXXFLAGS, CFLAGS.

I wrote user-config.jam manually and I'm able to build boost-exception.

**Comment by PhoebeHui** [comment]
AFAIK, user-config.jam only configured for windows build.
See https://github.com/microsoft/vcpkg/blob/master/ports/boost-modular-build-helper/boost-modular-build.cmake#L323

**Comment by Nemirtingas** [comment]
Are you sure ? CMake manual says that @ONLY will replace all variables between %%, it doesn't mean it works only on Windows.
Some variables are populated like 
@CURRENT_INSTALLED_DIR@
@VCPKG_PLATFORM_TOOLSET@

**Comment by PhoebeHui** [comment]
@Nemirtingas, sorry, I'm wrong, it also configured for linux and macOS via ports/boost-modular-build-helper/CMakeLists.txt, I missed that.

**Comment by Nemirtingas** [comment]
The configuration of user-config.jam works when I cross-compile for MacOS, maybe its some issue with clang on Windows ?

**Comment by Nemirtingas** [comment]
For Linux/Unix builds, vcpkg configures the user-config.jam at
https://github.com/microsoft/vcpkg/blob/master/ports/boost-modular-build-helper/boost-modular-build.cmake#L111
and
https://github.com/microsoft/vcpkg/blob/master/ports/boost-modular-build-helper/boost-modular-build.cmake#L76

This step is not called on Windows case because of
https://github.com/microsoft/vcpkg/blob/master/ports/boost-modular-build-helper/boost-modular-build.cmake#L103

I'm still investigating...

**Comment by Nemirtingas** [comment]
I successfully cross-compiled boost-thread and boost-exception, here are my changes:
https://github.com/Nemirtingas/vcpkg/commit/4ab844882c26d849b5d8ebda36511fec095246c4

Theses changes are not to be applied on vcpkg code as it will break boost compilation on Windows, but it can be used as a base idea to fix cross-compilation.

**Comment by Nemirtingas** [comment]
New vcpkg update added a new condition on windows compiler type: WIN32 or MSVC.
Fixed it here: https://github.com/Nemirtingas/vcpkg/commit/765ce91442fbc94b47d969e07d22bef9768d5d17

Maybe you'd like to add a condition on the compiler too so you setup boost's clang-win toolkit instead of msvc toolkit when using clang ?

**Comment by Nemirtingas** [comment]
Today's update broke again my fixes. Here are the new ones:
https://github.com/Nemirtingas/vcpkg/commit/420c0ae404f7df6b944797a10525bb45a99845d2

**Comment by Nemirtingas** [comment]
Fixed by #22274
